Complete third sprint

uncledeni opened the pull request:
Make public, authorized and not-authorized  routes;
Make pages 'Registration', 'SignIn', 'ResetPassword', 'ForgotPassword', 'Profile', 'Error';

Rework ingredient-details modal display

uncledeni committed:
Make layouts for several new pages; add basic routing

Add some router browser and routes to spp page

Create 'ActionBlock' universal component;
Complete 'Registration' page layout;
Complete 'SignIn' page layout;
Complete 'ResetPassword' page layout;
Complete 'ForgotPassword' page layout;

Start 'ProfileNavbar' component layout;
Start 'Profile' page layout;
Start 'Error' page layout;

uncledeni committed:
Minor visual updates and try/catch fix

Update 'ProfilePage' layout;
Update 'AppHeader' navbar visuals;
Update try/catch in 'burger-ingredients' and 'order-details' redux actions;

uncledeni committed:
Start importing auth into redux

Add new routes;
Update ProfileNavbar;
Start makeing 'auth' actions/reduces;

uncledeni committed:
Make 'Protected' component for routes overview;

uncledeni committed:
Update protected routes and auth check

make 'checkUserAuth' for authorisation confirm;

make 'fetchWithRefresh' wrapper;

make getUserData request;
update 'postOrder' request;

add 'setAuthChecked', 'getUser', 'setUser' additional functions;
update LoginThunk;
make SET_USER action/reducer;

incomplete:
~ logout
~ getUser in 'ProfilePage';

uncledeni committed:
Make some functional and visual updates

Fix 'logout';
Rename 'MainMenu' to 'Home';
Move 'Home' page to common routes;
Update 'ProfileNavbar';
Add controll buttons to 'Profile';
add patchUserData request;
Update 'auth' actions;
Update 'AppHeader';

uncledeni committed:
Make alternative routing for modal and ingredient page

Add routing for 'ingredinets/:id'; Create 'IngredientPage';
Werock routing access for 'reset-password' (need testing);
Add actions\reducers for 'getResetPasswordCode';

uncledeni committed:
Rerouting fixes and massive ingredients modal rework

Fix redirecting to 'reset-password' from 'forgot-password' and other routes; rework conditions in 'Protected';
Massive rework 'IngredientsDetails' modal:
    ~ remove component and render conditions from 'BurgerIngredients';
    ~ remove data update from 'IngredientPage';
    ~ implement render and data-update logic into 'IngredientDetails';
    ~ create modal wrapper for 'IngredientDetails';

minor fixes and updates

uncledeni committed:
Compete user patch and reset-password

Update navbars (app-header and profile);
Complete user data patch from profile;
Complete reset-password page;
clean auth actions'reducers;

gennady-bars commented on src/app/app.js:
```diff
@@ -1,8 +1,50 @@
+      </Routes>
+
+      {/* Show the modal when a `backgroundLocation` is set */}
+      {state?.backgroundLocation && (
```
Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars commented on src/pages/account/profile/profile.jsx:
```diff
@@ -0,0 +1,88 @@
+
+    return (
+        <div className={ProfileStyles.pageWrapper}>
+            <AppHeader />
```
` <AppHeader />` нужно поместить только в `App`, чтобы любая страница была с этим хэдером. Не нужно дублировать его в каждой странице

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -35,25 +33,16 @@ const IngredientsTabs = ({ current }) => {
 
     const dispatch = useDispatch();
     useEffect(() => {
         dispatch(getBurgerIngredients());
```
`dispatch(getBurgerIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/widgets/burger-ingredients/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -1,10 +1,32 @@
+    }, [dispatch])   
+
+    useEffect(() => {
+        dispatch(getBurgerIngredients());
```
`dispatch(getBurgerIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -90,7 +76,10 @@ const IngredientsStack = (props) => {
+                        </Link>
                     ))}
                 </div>
             </div>
```
`IngredientDetails.propTypes` нужно удалить, так как в пропсы ничего не приходит теперь

gennady-bars commented on src/shared/utils/protected-route.jsx:
```diff
@@ -0,0 +1,32 @@
+        return null;
+    }
+
+    if (onlyUnAuth && user) {
+        const { from } = location.state || { from: { pathname: '/' } };
+        return <Navigate to={from} />;
+    }
+
+    if (!onlyUnAuth && !user) {
+        return <Navigate to='/login' state={{ from: location }} />;
+    }
+
+    if (window.location.pathname === '/reset-password' && location.state === null) {
+        return <Navigate to='/login' state={{ from: location }} />;
```
Хорошо сделан `Protected`

gennady-bars commented on src/pages/registration/sign-in/sign-in.jsx:
```diff
@@ -0,0 +1,55 @@
+
+export const SignInPage = () => {
+    const dispatch = useDispatch();
+    const [email, setEmail] = useState('');
+    const onChangeEmail = e => {
+        setEmail(e.target.value)
+    }
+
+    const [password, setPassword] = useState('');
+    const onChangePassword = e => {
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/registration/sign-in/sign-in.jsx:
```diff
@@ -0,0 +1,55 @@
+            <main className={SignInStyles.mainWrapper}>
+                <div className={SignInStyles.mainContainer}>
+                    <h1 className={`${SignInStyles.title} text text_type_main-medium`}>Вход</h1>
+                    <EmailInput
```
Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.

gennady-bars commented on src/pages/registration/sign-in/sign-in.jsx:
```diff
@@ -0,0 +1,55 @@
+                        <Button htmlType="button"
+                        type="primary"
+                        size="large"
+                        onClick={() => dispatch(LoginThunk({email, password}))}>
```
**Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**

gennady-bars commented on src/shared/services/actions/auth.js:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+export const LoginThunk = (data) => {
```
`LoginThunk`

Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно
* Хорошо сделан `Protected`

но есть некоторые недочеты:



* ` <AppHeader />` нужно поместить только в `App`, чтобы любая страница была с этим хэдером. Не нужно дублировать его в каждой странице
* Скрин https://disk.yandex.ru/i/f7AM_LOPdhMhRA Неавторизованный пользователь **не должен** отправлять заказ на сервер. При клике на кнопку его нужно переводить на логин и после удачного логина обратно к заказу.
* `dispatch(getBurgerIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* `IngredientDetails.propTypes` нужно удалить, так как в пропсы ничего не приходит теперь
* Если есть **инпуты**, то они должны быть внутри тега `form`, и на этот тег `form` нужно навешивать обработчик `onSubmit`, а не обработчик `onClick` на кнопку сабмита.
* **Обработчик сабмита** нужно навешивать только на тег `form` с событием `submit`, а не на кнопку сабмита с событием `click`, так как сабмит формы происходит ещё при нажатии `Enter`, и он не будет работать, если навесить обработчик клика на кнопку только. **Это нужно исправить везде, где есть инпуты и форма**
* Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта


**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

uncledeni committed:
Pull Request fixes for third sprint

Move 'AppHeader' to 'App' level; move getBurgerIngredients to 'App' level;
Rework all forms 'button-onClick' to submit;
Update all requests to server from forms;
Make hook useForm for form data collecting;
Make 'setDefaultValues' function for reset data in form on 'Profile' page in 'Cancel' case;
Make wrapper for order request for unauth users;
Delete 'IngredientDetails.propTypes';

make some minor updates

uncledeni commented:
Основные изменение внесены
Лоадер постараюсь сделать в течении недели

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
